p2p/src/simulated/network.rs

```diff
@@ -500,39 +502,106 @@ impl crate::Receiver for Receiver {
     }
 }
 
-// A peer in the simulated network.
-// The peer can receive messages over any registered channel.
+/// A peer in the simulated network.
+///
+/// The peer can register channels, which allows it to receive messages sent to the channel from other peers.
 struct Peer {
     // Socket address that the peer is listening on
     socket: SocketAddr,
 
-    // Map from channel to a sender that can deliver messages to that channel
-    mailboxes: Arc<Mutex<HashMap<Channel, mpsc::UnboundedSender<Message>>>>,
+    // Control to register new channels
+    control: mpsc::UnboundedSender<(Channel, oneshot::Sender<MessageReceiverResult>)>,
 }
 
 impl Peer {
+    /// Create and return a new peer.
+    ///
+    /// The peer will listen for incoming connections on the given `socket` address.
+    /// `max_size` is the maximum size of a message that can be sent to the peer.
     fn new<E: Spawner + RNetwork<Listener, Sink, Stream>>(
         runtime: &mut E,
         public_key: PublicKey,
         socket: SocketAddr,
         max_size: usize,
     ) -> Self {
-        let result = Self {
-            socket,
-            mailboxes: Arc::new(Mutex::new(HashMap::new())),
-        };
+        // The control is used to register channels.
+        // There is exactly one mailbox created for each channel that the peer is registered for.
+        let (control_sender, mut control_receiver) = mpsc::unbounded();
+
+        // Whenever a message is received from a peer, it is placed in the inbox.
+        // The router polls the inbox and forwards the message to the appropriate mailbox.
+        let (inbox_sender, mut inbox_receiver) = mpsc::unbounded();
+
+        // Spawn router
+        runtime.spawn("router", async move {
+            // Map of channels to mailboxes (senders to particular channels)
+            let mut mailboxes = HashMap::new();
+
+            // Continually listen for control messages and outbound messages
+            loop {
+                select! {
```

Comment:
Just to reiterate, the `select!` macro is "order-biased". So make sure to think carefully about the order of each item below.